[Project1/배건호] Notion 클로닝 프로젝트

meowTarae opened the pull request:
## 📌 과제 설명 

바닐라 JS만을 이용한 노션 클로닝

## 👩‍💻 요구 사항과 구현 내용 

### 기본 요구 사항
- [x] 글 단위를 Document라고 합니다. Document는 Document 여러개를 포함할 수 있습니다.

- [x] 화면 좌측에 Root Documents를 불러오는 API를 통해 루트 Documents를 렌더링합니다.

  - [x] Root Document를 클릭하면 오른쪽 편집기 영역에 해당 Document의 Content를 렌더링합니다.

  - [x] 해당 Root Document에 하위 Document가 있는 경우, 해당 Document 아래에 트리 형태로 렌더링 합니다.

  - [x] Document Tree에서 각 Document 우측에는 + 버튼이 있습니다. 해당 버튼을 클릭하면, 클릭한 Document의 하위 Document로 새 Document를 생성하고 편집화면으로 넘깁니다.

- [x] 편집기에는 기본적으로 저장 버튼이 없습니다. Document Save API를 이용해 지속적으로 서버에 저장되도록 합니다.

- [x] History API를 이용해 SPA 형태로 만듭니다.

  - [x] 루트 URL 접속 시엔 별다른 편집기 선택이 안 된 상태입니다.

  - [x] /documents/{documentId} 로 접속시, 해당 Document 의 content를 불러와 편집기에 로딩합니다.

  <br>

### 보너스 요구 사항
 - [ ] 기본적으로 편집기는 textarea 기반으로 단순한 텍스트 편집기로 시작하되, 여력이 되면 div와 contentEditable을 조합해서 좀 더 Rich한 에디터를 만들어봅니다.

- [ ] 편집기 최하단에는 현재 편집 중인 Document의 하위 Document 링크를 렌더링하도록 추가합니다.

- [ ] 편집기 내에서 다른 Document name을 적은 경우, 자동으로 해당 Document의 편집 페이지로 이동하는 링크를 거는 기능을 추가합니다.

- [ ] 그외 개선하거나 구현했으면 좋겠다는 부분이 있으면 적극적으로 구현해봅니다!

## ✅ 피드백 반영사항  <!-- 지난 코드리뷰에서 고친 사항을 적어주세요. 재PR 시에만 사용해 주세요! (재PR 아닌 경우 삭제) -->

## ✅ PR 포인트 & 궁금한 점

초반에 구현에 대한 갈피를 잡지 못하여 기본 요구사항만 겨우 충족하였습니다..

이벤트 리스너 함수를 리팩토링 하는 방법이 궁금합니다.

데이터 흐름과 올바르지 못한 코드 등 피드백 주시면 감사하겠습니다

meowTarae committed:
🎉 Initial Commit

meowTarae committed:
Feat: 기본 틀 생성

meowTarae committed:
노션

Jae0o commented on src/sidebar/Sidebar.js:
```diff
@@ -0,0 +1,39 @@
+  const $sidebar = document.createElement("aside");
+  $sidebar.className = "sidebar";
+
+  // 헤더를 만들건데, 만약 로컬스토리지에 먼가 있따? 바로 삭제.
+  new Title({
+    $target: $sidebar,
+  });
+
+  new Header({
+    $target: $sidebar,
+    text: "개인 페이지",
+  });
```
( P5 ) 제가 주석의 의미를 파악하지 못해서 그럴수도 있긴하지만,
Header.js Component 안에 header tag를 사용하셨는데, 보통 header 태그를 사용할때 태그 내부에
제목 , 로고, icon , 저자 정보 등등.. 이 들어 간다고 하더라구요!
그래서 별도로 제작하신 Title.js 는 해당 App의 제목을 의미하는것 같아서 header 태그 안에 같이 포함됐으면 어땟을까 라는
의미론적 의견을 드려봅니다! 🧐

meowTarae replied:
오호! header에 여러가지를 담는 줄은 몰랐네요.
덕분에 좋은 지식 알게되었습니다 👍

Jae0o reviewed (commented):
건호님.. 최대한 좋은 리뷰를 남겨드리고싶었는데 빈틈이 없으시네요.. 🥲
항상 변수처리나 코드 간결성? 에서 너무 많이 배워갑니다. 그래서 이번 리뷰에서 태그의 의미 부분에서 많이 
남겨드린 것 같아서 아쉽네요..
이번에도 역시 잘 배워갑니다! 👍

hellosonic-r commented on src/sidebar/components/elements.js:
```diff
@@ -0,0 +1,38 @@
+export function ToggleElements({ id, isWrap }, depth) {
+  return `
+    <i class="toggle fas ${
+      isWrap ? "fa-chevron-right" : "fa-chevron-down"
+    }" data-id='${id}' data-isWrap='${isWrap}' style="padding-left: ${
+    depth * 8
+  }px;">
+    </i>    
+  `;
+}
+
+export function LiElements({ id, title }) {
+  return `
+    <li class='title' data-id='${id}' > 
+      ${title} 
+    </li>
+  `;
+}
+
+export function AddElements({ id }) {
+  return `
+    <i class="fas fa-plus add" data-id='${id}' ></i>
+  `;
+}
+
+export function removeElements({ id }) {
+  return `
+    <i class="fas fa-trash-alt remove" data-id='${id}' ></i>
+  `;
+}
+
+export function NoPageElements({ depth }) {
+  return `
+    <span class='noPage' style="padding-left: ${depth * 2}px; cursor: default;">
+      하위 페이지 없음
+    </span>
+  `;
+}
```
각 기능을 가지는 요소들을 따로 js파일로 분리해서 작성하신게 너무 좋았던 것 같습니다👍
덕분에 Documents.js 안의 코드를 이해하기 편했던 것 같아요👍

hellosonic-r commented on src/sidebar/Documents.js:
```diff
@@ -0,0 +1,130 @@
+          ${
+            documents.length
+              ? documents.map((document) => paint(document, depth + 1)).join("")
+              : NoPageElements({ depth })
```
저는 미처 생각하지 못한 부분인데, 하위 documents가 없을 시 따로 표시해주는 부분이 코드가 너무 깔끔하게 작성 잘 되어있는 것 같아요. 건호님 코드 볼 때마다 느끼는 거지만 가독성 좋게 작성되어있어서 보기 편한 것 같습니다! 💯

hellosonic-r reviewed (commented):
건호님 과제하신거 잘 보구갑니다! 
역시나 이번 과제 코드도 가독성 좋게 잘 작성해주셨네요! 💯
과제할 때 놓쳤던 부분을 건호님 코드 보고 저도 아이디어를 얻었네용 ㅎㅎ 감사합니다👍

wldlsgur reviewed (commented):
건호님 Notion 프로젝트하신다고 고생 많으셨어요!!!

중간에 생각처럼 잘 안되서 많이 힘드신거 같았는데 코드 보니깐 대처 능력도 뛰어나신 것 같아요!!ㅎㅎ 재영님 말씀처럼 건호님 코드가 탄탄해서 제가 감히 콕 찝을 부분이 많이 없네요 ㅜㅜ

그래도 더 열심히 노력해서 최대한 적극적으로 리뷰 작성하도록 노력하겠습니다!!

joonwonBaek commented on src/utils/api.js:
```diff
@@ -0,0 +1,58 @@
+  }
+};
+
+// delete
+export const deleteDocument = async (id) => {
+  await request(`/documents/${id}`, {
+    method: "DELETE",
+  });
+};
+
+// post
+export const insertDocument = async (data) => {
+  const res = await request("/documents", {
+    method: "POST",
+    body: JSON.stringify(data),
+  });
+
+  return res;
+};
+
+// get
+export const getDocument = async (id) => {
+  const res = await request(`/documents/${id}`, {
+    method: "GET",
+  });
+
+  return res;
+};
+
+// put
+export const updateDocument = async ({ id, title, content }) => {
+  const res = await request(`/documents/${id}`, {
+    method: "PUT",
+    body: JSON.stringify({
+      title,
+      content,
+    }),
+  });
```
request를 api.js에 받으니 더 보기 깔끔한거 같습니다 저도 리팩토링할 때 참고해보겠습니다👍

joonwonBaek reviewed (commented):
건호님 프로젝트 너무 고생하셨습니다! 코드 리뷰를 하는데 강의에서 진행한 실습 코드 분석을 정말 완벽하고 열심히 하셨다는 생각이 들었습니다. 그리고 너무 깔끔한거 같습니다👍

Changyu-Ryou reviewed (commented):
건호님! 과제하시느라 고생 많으셨어요! 

추가 요구사항은 구현하지 못하셨다고 하셨는데 코드 읽으면서 기본 요구사항 구현하실 때 많이 고민하시고 꼼꼼하게 잘 구현해주신 것 같아서 좋았습니다! 👍👍👍 

정말 고생 많으셨어요!! 질문주신 내용은 코멘트에 달아두었어요~!

meowTarae commented on src/sidebar/Documents.js:
```diff
@@ -0,0 +1,130 @@
+    this.render();
+  };
+
+  const paint = ({ id, title, documents }, depth) => {
+    const isWrap = getItem(`isWrap-${id}`, true);
+    return `
+      <div class='documents'>
+        <div class='document' data-id='${id}'>
+          ${ToggleElements({ id, isWrap }, depth)}
+          ${LiElements({ id, title })}
+          ${AddElements({ id })}
+          ${removeElements({ id })}
+        </div>  
+        <ul style="display: ${isWrap ? "none" : "block"};">
+          ${
+            documents.length
+              ? documents.map((document) => paint(document, depth + 1)).join("")
+              : NoPageElements({ depth })
+          }
+        </ul>
+      </div>
+    `;
+  };
+
+  this.render = () => {
+    if (this.state.length > 0) {
+      $documents.innerHTML = `
+       <ul class='top'>
+         ${this.state.map((document) => paint(document, 0)).join("")}
+       </ul>
+      `;
+    } else {
+      $documents.innerHTML = `
+        <h4>페이지를 추가해 주세요</h4>
+      `;
+    }
+  };
```
두 분 좋은 의견 감사드립니다!
저도 ul 태그에는 li 태그가 들어가는 것이 좋다고 생각됩니다 👍 
허나 li 태그 삽입 시, 기존의 정렬이 모두 무너져내려서.. ul, li 태그 대신 div, span 태그로 수정하였습니다.

meowTarae committed:
🎨 Improve: '개선 태그 네이밍'

ul 태그 내부에는 1개 이상의 li 태그가 있는 것이 좋음.

다만, li 태그 사용 시, 기존의 css 구조가 무너짐.

ul 태그 -> div 태그
li 태그 -> div 태그 및 span 태그

위와 같이 개선

meowTarae commented on src/sidebar/Header.js:
```diff
@@ -0,0 +1,29 @@
+
+  this.render();
+
+  $header.addEventListener("click", (e) => {
+    const { target } = e;
+
+    if (target.classList.contains("add")) {
+      const isExist = getItem(CREATED_DOCUMENTS_PARENT_ID_KEY, null);
+      if (isExist) removeItem(CREATED_DOCUMENTS_PARENT_ID_KEY);
+      push(`/documents/new`);
+    }
+  });
```
if문 중첩을 줄일 수 있는 좋은 방법이네요! 감사합니다.

meowTarae committed:
♻️ Refactor: 'click 이벤트 리스너 함수 리팩토링'

meowTarae committed:
♻️ Refactor: 'header의 의미론적 관점에 맞게 코드 재배치'

header 태그에는 제목, 로고, icon, 저자 정보 등이 들어간다.

따라서 Title을 별도로 분리하는 것이 아닌, header 태그에 넣어 두는 것이 의미론적 관점에서 조금 더 좋아보인다.

때문에 Title.js 코드를 Header.js 로 이전하였다.

meowTarae commented on src/sidebar/components/elements.js:
```diff
@@ -0,0 +1,38 @@
+export function ToggleElements({ id, isWrap }, depth) {
```
클릭되는 부분이니까 버튼 태그를 사용하는 것이 맞는 것 같네요!

재영님 덕분에 앞으로는 태그를 조금 더 상황에 맞게 사용할 수 있게끔 신경쓰게 될 것 같습니다~ 감사합니다!

meowTarae committed:
🐛 Fix: '수정 import error'

Title.js 파일 삭제 후 지우지않았던 Title.js 임포트 문을 지워줌으로써 에러를 해결.

meowTarae commented on src/utils/api.js:
```diff
@@ -0,0 +1,58 @@
+        "Content-Type": "application/json",
+      },
+    });
+
+    if (res.ok) {
+      return await res.json();
+    }
```
400~500번대 각각의 status가 이렇게 다른 성질을 지니고 있는 줄 몰랐습니다.
덕분에 새로 알게 되었습니다 😄

말씀주신 것 처럼, 현재의 코드에 맞게.. 로그인이 구현되어 있다면 401, 403 status를 활용하고, 현재는 404 status의 경우를 추가해보았습니다.

좋은 정보 감사합니다.

meowTarae committed:
✨ Feat: '예외 처리 추가'

meowTarae committed:
🔥 Remove: '쓸데없는 주석'

meowTarae commented on src/editor/Editor.js:
```diff
@@ -0,0 +1,50 @@
+
+  this.state = initialState;
+
+  this.setState = (nextState) => {
+    this.state = nextState;
+    $editor.querySelector("[name=title]").value = this.state.title;
+    $editor.querySelector("[name=content]").value = this.state.content;
+    this.render();
+  };
+
+  this.render = () => {
+    if (!isInit) {
+      $editor.innerHTML = `
+          <div class='editor'>
+            <input class='editorTitle' placeholder="제목을 입력해 주세요" type="text" name='title' value="${this.state.title}" />
+            <textarea class='editorContent' placeholder="내용을 입력해 주세요" name="content"  >${this.state.content}</textarea>
+          </div>
+        `;
+      isInit = true;
+    }
+  };
```
좋은 의견 감사드립니다 👍

meowTarae commented on src/sidebar/Documents.js:
```diff
@@ -0,0 +1,130 @@
+          ${AddElements({ id })}
+          ${removeElements({ id })}
+        </div>  
+        <ul style="display: ${isWrap ? "none" : "block"};">
```
말씀처럼 별도의 hidden 클래스를 만드는 등의 방법으로 관리한다면 css 파트와 js 파트를 구분하는 것은 좋은 방법인 것 같습니다 !

meowTarae commented on src/sidebar/Documents.js:
```diff
@@ -0,0 +1,130 @@
+      }
+      deleteDocument(id);
+      onDeleteDocument();
+      // 이거 말고 sidebar를 리렌더하자
+      push(`/`);
```
문서를 삭제했을 때 사이드바에 있는 문서들 리스트를 갱신하기 위함이었습니다.

meowTarae committed:
♻️ Refactor: 'url 추출 코드를 정규표현식으로'

meowTarae commented on src/App.js:
```diff
@@ -1,7 +1,43 @@
+    if (pathname === "/") {
+      removeItem(CREATED_DOCUMENTS_PARENT_ID_KEY);
+    }
+    if (pathname.indexOf("/documents/") === 0) {
+      const [, , documentId] = pathname.split("/");
```
정규표현식이 확실히 더 안전하고 깔끔해보입니다 !

meowTarae commented on index.html:
```diff
@@ -1,9 +1,17 @@
-</html>
\ No newline at end of file
+<!DOCTYPE html>
+<html lang="kr">
```
앗..! kr이 아니라 ko 였군요! 😅

meowTarae committed:
🩹 Fix: 'html lang: kr -> ko'

meowTarae commented on src/utils/storage.js:
```diff
@@ -0,0 +1,18 @@
+const storage = window.localStorage;
+
+export const getItem = (key, defaultValue) => {
+  try {
+    const storedValue = storage.getItem(key);
+    return storedValue ? JSON.parse(storedValue) : defaultValue;
+  } catch (e) {
+    return defaultValue;
+  }
+};
+
+export const setItem = (key, value) => {
+  storage.setItem(key, JSON.stringify(value));
+};
+
+export const removeItem = (key) => {
+  storage.removeItem(key);
+};
```
싱글톤 패턴이 무엇인지 잘 몰라 검색해보니, 하나의 클래스를 선언하여 해당 클래스의 프로퍼티나 메서드를 사용하는 느낌이었습니다.

JS예시를 찾아보진 못했지만, 하나의 request 라는 객체 안에 get, post 등의 함수를 집어넣어 사용하면, 코드가 더 간결해지고 깔끔해져서 좋겠다고 생각했습니다.

덕분에 싱글톤 패턴에 대하여 알게되었습니다! 감사합니다 👍

meowTarae commented on src/utils/api.js:
```diff
@@ -0,0 +1,58 @@
+    const res = await fetch(`${API_END_POINT}${url}`, {
+      ...options,
+      headers: {
+        "x-username": "meowTarae",
```
앗..! 고친다고 해놓고 깜빡했네요 😅

meowTarae committed:
♻️ Refactor: '상수 값 리팩토링'

meowTarae committed:
🐛 Fix: '.js 미기입 수정'

meowTarae committed:
🔥 Remove: '주석 및 미사용 import'
